### PR TITLE
threshold monitoring for beacon processes

### PR DESCRIPTION
**Pull request by CluEleSsUK**
beacon processes now log some error and warning messages whenever we get close to or cross a threshold number of failures per node

### CODE REVIEW

File: metrics/threshold_monitor.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+					t.log.Errorw(
+						"failed connections crossed threshold in the last minute",
+						"beaconID", t.beaconID,
+						"threshold", t.threshold,
+						"failures", len(failingNodes),
+						"nodes", strings.Join(failingNodes, ","),
+					)
+				} else if len(failingNodes) >= t.threshold/2 {
+					t.log.Warnw(
+						"failed connections crossed half threshold in the last minute",
```
Comment (AnomalRoil) [inline metrics/threshold_monitor.go:64]: Same. Looks more like period given the sleep below to me

File: chain/beacon/node.go [hunk 2]
```diff
@@ -87,16 +90,17 @@ func NewHandler(c net.ProtocolClient, s chain.Store, conf *Config, l log.Logger,
...
+		client:           c,
+		crypto:           v,
+		chain:            store,
+		ticker:           ticker,
+		addr:             addr,
+		ctx:              ctx,
+		ctxCancel:        ctxCancel,
+		l:                l,
+		version:          version,
+		thresholdMonitor: dmetrics.NewThresholdMonitor(conf.Group.ID, l, conf.Group.Threshold),
```
[the remark this answers is not in the document]
Comment (CluEleSsUK) [reply chain/beacon/node.go:103]: I initially had it in the process, but it was essentially just being passed into the Handler. Also when resharing happens, we create a new handler but not a new process, so it fit the lifecycle to recreate it with the handler

File: metrics/threshold_monitor.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+			default:
+				t.lock.RLock()
+				var failingNodes []string
+				for address := range t.failedConnections {
+					failingNodes = append(failingNodes, address)
+				}
+
+				if len(failingNodes) >= t.threshold {
+					t.log.Errorw(
+						"failed connections crossed threshold in the last minute",
```
[the remark this answers is not in the document]
Comment (CluEleSsUK) [reply metrics/threshold_monitor.go:55]: true, it's technically the last $whateverPeriodWePassedIn, although by default it's 1min

File: metrics/threshold_monitor.go (hunk 1 above)
Comment (CluEleSsUK) [reply metrics/threshold_monitor.go:63]: the period here is different to the network period, by default it's 1min

**Review by AnomalRoil — APPROVED** [review]

**Pull request merged**